src/extensionManager.ts

```diff
@@ -0,0 +1,189 @@
+    )
+
+    // Create our own console
+    const channel = vscode.window.createOutputChannel(`Jest (${workspaceFolder.name})`)
+
+    const failDiagnostics = vscode.languages.createDiagnosticCollection(`Jest (${workspaceFolder.name})`)
+
+    this.extByWorkspace[workspaceFolder.name] = new JestExt(
+      this.context,
+      workspaceFolder,
+      jestWorkspace,
+      channel,
+      pluginSettings,
+      this.debugCodeLensProvider,
+      this.debugConfigurationProvider,
+      failDiagnostics,
+      instanceSettings
+    )
+  }
+  unregister(workspaceFolder: vscode.WorkspaceFolder) {
+    this.unregisterByName(workspaceFolder.name)
+  }
+  unregisterByName(name: string) {
+    const extension = this.extByWorkspace[name]
+    if (extension) {
+      extension.deactivate()
+      delete this.extByWorkspace[name]
+    }
+  }
+  unregisterAll() {
+    Object.keys(this.extByWorkspace).forEach(key => this.unregisterByName(key))
+  }
+  shouldStart(workspaceFolderName: string): boolean {
+    const { commonPluginSettings: { enabledWorkspaceFolders, disabledWorkspaceFolders } } = this
+    return (
+      // start if enabledWorkspaceFolders contains current workspace
+      enabledWorkspaceFolders.includes(workspaceFolderName) ||
+      // dont start if disabledWorkspaceFolders doesn't contain current workspace
+      !disabledWorkspaceFolders.includes(workspaceFolderName)
+    )
+  }
+  getByName(workspaceFolderName: string) {
+    return this.extByWorkspace[workspaceFolderName]
+  }
+  getByDocUri(uri: vscode.Uri) {
+    const workspace = vscode.workspace.getWorkspaceFolder(uri)
+    if (workspace) {
+      return this.getByName(workspace.name)
+    }
+  }
+  async get() {
+    const workspace =
+      vscode.workspace.workspaceFolders.length <= 1
+        ? vscode.workspace.workspaceFolders[0]
+        : await vscode.window.showWorkspaceFolderPick()
+
+    const instance = workspace && this.getByName(workspace.name)
+    if (instance) {
+      return instance
+    } else if (workspace) {
+      throw new Error(`No Jest instance in ${workspace.name} workspace`)
+    }
+  }
+  registerCommand(command: string, callback: (extension: JestExt, ...args: any[]) => any, thisArg?: any) {
+    return vscode.commands.registerCommand(command, async (...args) => {
+      const extension = await this.get()
+      if (extension) {
+        callback.call(thisArg, extension, ...args)
+      }
+    })
+  }
+  onDidChangeConfiguration(e: vscode.ConfigurationChangeEvent) {
+    if (e.affectsConfiguration('jest')) {
+      const updatedSettings = getExtensionWindowSettings()
+      this.debugCodeLensProvider.showWhenTestStateIn = updatedSettings.debugCodeLens.enabled
+        ? updatedSettings.debugCodeLens.showWhenTestStateIn
+        : []
+    }
+    vscode.workspace.workspaceFolders.forEach(workspaceFolder => {
+      const jestExt = this.getByName(workspaceFolder.name)
+      if (jestExt && e.affectsConfiguration('jest', workspaceFolder.uri)) {
```

Comment:
should we register/unregister workspace when settings such as `jest.disabledWorkspaceFolders` change? Looks like the `jestExt` was not changed when I add/remove the workspace via the setting`jest.disabledWorkspaceFolders`